WebDriverAgentLib/Categories/XCUIElement+FBFind.m

```diff
@@ -83,18 +85,42 @@ - (void)descendantsWithProperty:(NSString *)property value:(NSString *)value par
   // XPath will try to match elements only class name, so requesting elements by XCUIElementTypeAny will not work. We should use '*' instead.
   xpathQuery = [xpathQuery stringByReplacingOccurrencesOfString:@"XCUIElementTypeAny" withString:@"*"];
   NSArray *matchingSnapshots = [self.lastSnapshot fb_descendantsMatchingXPathQuery:xpathQuery];
-  NSArray *allElements = [[self descendantsMatchingType:XCUIElementTypeAny] allElementsBoundByIndex];
-  NSArray *matchingElements = [self filterElements:allElements matchingSnapshots:matchingSnapshots];
+  if (nil == matchingSnapshots) {
+    return @[];
+  }
+  // Prefiltering elements speeds up search by XPath a lot, because [element resolve] is the most expensive operation here
+  NSDictionary *elementsByType = [self filterPotentialMatchesByType:matchingSnapshots];
+  NSArray *matchingElements = [self filterElements:elementsByType matchingSnapshots:matchingSnapshots];
   return matchingElements;
 }
 
-- (NSArray<XCUIElement *> *)filterElements:(NSArray<XCUIElement *> *)elements matchingSnapshots:(NSArray<XCElementSnapshot *> *)snapshots
+- (NSDictionary<NSNumber *, NSArray<XCElementSnapshot *> *> *)filterPotentialMatchesByType:(NSArray<XCElementSnapshot *> *)snapshots
+{
+  NSMutableSet *matchingTypes = [NSMutableSet set];
+  [snapshots enumerateObjectsUsingBlock:^(XCElementSnapshot *snapshot, NSUInteger snapshotIdx, BOOL *stopSnapshotEnum) {
+    [matchingTypes addObject:@(snapshot.elementType)];
+  }];
+  NSMutableDictionary *result = [NSMutableDictionary dictionary];
+  [matchingTypes enumerateObjectsUsingBlock:^(NSNumber *elementTypeAsNumber, BOOL *stopEnum) {
+    XCUIElementType elementType = (XCUIElementType)elementTypeAsNumber.unsignedIntegerValue;
+    NSArray *descendantsOfType = [[self descendantsMatchingType:elementType] allElementsBoundByIndex];
+    result[elementTypeAsNumber] = descendantsOfType;
+  }];
+  return result.copy;
+}
+
+- (NSArray<XCUIElement *> *)filterElements:(NSDictionary<NSNumber *, NSArray<XCElementSnapshot *> *> *)elementsMap matchingSnapshots:(NSArray<XCElementSnapshot *> *)snapshots
 {
   NSMutableArray *matchingElements = [NSMutableArray array];
   [snapshots enumerateObjectsUsingBlock:^(XCElementSnapshot *snapshot, NSUInteger snapshotIdx, BOOL *stopSnapshotEnum) {
+    NSArray *elements = elementsMap[@(snapshot.elementType)];
     [elements enumerateObjectsUsingBlock:^(XCUIElement *element, NSUInteger elementIdx, BOOL *stopElementEnum) {
-      [element resolve];
-      if ([element.lastSnapshot _matchesElement:snapshot]) {
+      id lastSnapshot = [element lastSnapshot];
```

Comment:
Why `id` here?
Can you add `fb_lastsnapshot` method to `XCUIElement` helper category that does that? We do same thing in few places.